disabled web app resource from being published

hyi opened the pull request:
This PR addressed the issue #1648 to disable web app resource from being published. This is important since we don't want people to wrongly publish web app resources. This needs to go to AGU if possible. @pkdash Can you give a quick review at your earliest convenience?

pkdash commented on hs_core/models.py:
```diff
@@ -1686,6 +1686,12 @@ def size(self):
 
+    @property
+    def can_be_published(self):
+        if self.can_be_public_or_discoverable() and self.resource_type != 'ToolResource':
```
I think it would be better to override this property in the ToolResource app and not do this check ```self.resource_type != 'ToolResource'``` here.

pkdash commented on hs_core/hydroshare/resource.py:
```diff
@@ -1123,10 +1123,12 @@ def publish_resource(user, pk):
-        raise ValidationError("This resource cannot be published since it does not have required metadata or content files.")
+    if not resource.can_be_published():
+        raise ValidationError("This resource cannot be published since it does not have required "
+                              "metadata or content files or it is a Web App resource.")
```
I would change the message ```or it is a Web App resource``` to something like ```or this resource type is not allowed for publication.``` so that we don't have to hard code the resource type name.

hyi commented:
@pkdash Thanks for the prompt code review. I have addressed your code review comments so far. See if you can +1 now or have any further comments to be addressed. Hope this PR can go into AGU. Thanks.

pkdash commented on theme/templates/resource-landing-page/top-right-buttons.html:
```diff
@@ -50,7 +50,7 @@
 
-                        {% if page.perms.delete and cm.can_be_public_or_discoverable %}
+                        {% if page.perms.delete and cm.can_be_published %}
                             {% if cm.resource_type|lower == "collectionresource" and cm.are_all_contained_resources_published or cm.resource_type|lower != "collectionresource" %}
```
The CollectionResource should override the ```can_be_published``` property so that we do not have to check this specifically for CollectionResource here.

pkdash replied:
Since you have written can_be_published in CollectionResource now, we don't need the check in line#54 right?

pkdash commented on hs_core/models.py:
```diff
@@ -1688,7 +1688,12 @@ def verbose_name(self):
+        publication such as the Web App resource
+        :return:
+        """
+        if self.can_be_public_or_discoverable():
```
Better would be this one liner: ```return self.can_be_public_or_discoverable()```

hyi commented:
@pkdash Have addressed your further comments. See it if looks good to you now. Thanks.

hyi commented:
@pkdash Sorry I forgot to add that updated file in the last commit in which I have removed that extra check for collection resource type. Just added it now. However, I did discover one issue I am puzzled by. It appears I cannot call a property ```can_be_public_or_discoverable()``` directly within another property ```can_be_published()```. It does not raise error, but does not seem to work as it should. I am currently trying to figure that one out.

hyi commented:
@pkdash Have fixed that property-calling-property issue. I need to remove the pair of parenthesis for calling a property. Now everything seems to work as it should. See if you can give a +1 now. Thanks.

pkdash commented:
+1 Merge after Jenkins checks are successful.
